### ISSUE TITLE
Switch base distribution to Ubuntu 16.04 LTS Xenial Xerus

### ISSUE THREAD

**Issue by almereyda** [mail-in-a-box/mailinabox#758]
[The preflight](https://github.com/mail-in-a-box/mailinabox/blob/65add24e2abcc0015d7a49645218e13e37e7d5b6/setup/preflight.sh#L10-L12) issues a hard check for the used Ubuntu distribution.

With `16.04` arriving within a month and already having a beta.

Are there plans to start a new branch `xx` for Xenial Xerus to prepare for the upcoming migrations?

Häppy Hacking!

**Comment by JoshData** [comment]
Ugh has it been two years already?

I'm tempted to wait two more years and switch at what would presumably be 18.04 LTS....

**Comment by almereyda** [comment]
I believe this is due to too much maintenance needed for MIAB either way?

How could the community help out with a port / fork / migration path?

**Comment by JoshData** [comment]
So the first question is whether we want to force all existing users to kill their current boxes and set up a new one now or in two years? I don't want to maintain two branches, so once the next branch is released there won't be any further development on the old one (except critical security issues for some limited amount of time).

**Comment by dhpiggott** [comment]
Here's another idea, which may not be worth anything but is worth mentioning on the off chance it is: what about having only one branch, but implementing the support for 16.04 by adding conditionals in each place that 16.04 has to be treated differently to 14.04? This could be maintained for e.g. 9 months, giving users plenty of time to upgrade. At the end of that period, the conditionals could be removed and the scripts would revert to supporting only one Ubuntu version. I guess the feasibility of this would depend on how many things need to be changed for 16.04, and only experimentation will confirm that.

**Comment by ianberinger** [comment]
I think what @dhpiggott is suggesting could work very well, as the main change with 16.04 is the usage of systemd. I've been running my personal installation on 15.10 for a while now and I had to change only a few things to get it working. The only thing I couldn't get working was the [dovecot-lucene](http://packages.ubuntu.com/search?suite=default&section=all&arch=any&keywords=dovecot-lucene&searchon=names) package. I certainly could do a PR once I've cleaned up my code.

**Comment by JoshData** [comment]
If it's that easy, yes, that would work.

**Comment by DamonGant** [comment]
In my experience the live LTS updater is pretty unstable, however, I think it's worth a try for - as you often emphasize - unmodified boxes to do an in-place upgrade along with mail in a box.

**Comment by JoshData** [comment]
I've decided that we need a complete test suite before we change the base Ubuntu version (or anything more drastic). Otherwise we'll have no idea if the system is working properly on the new version. See #777. Thanks.

**Comment by ponychicken** [comment]
Maybe it's worth considering something like NixOS.
It is completely based around a declarative configuration.
You specify what services you want and how they should be configured,
and it takes care of building everything.

http://nixos.org/nixos/about.html

**Comment by DamonGant** [comment]
The problem with that is that it it stops being easy to use. Most cloud providers don't have readily-available images for NixOS.

Customization, while discouraged, would also be harder; a lot of people consider dpkg the one Linux package manager and would probably be left confused.

Not to mention the entire thing would need to be ported to an environment that is probably unknown to MiaB developers.

**Comment by JoshData** [comment]
Just a thought to reconsider shipping as a .deb if we ever make it to 16.04.  #183

**Comment by anoma** [comment]
If you are happy to tie your self to Ubuntu then I would use snappy packages rather than deb for three reasons:

# in theory they let you roll back (something a .deb will never do)
# containerization like security
# but mainly you have far more control over the publishing process and you will not need to maintain your own repo or be beholden on official repo maintainers for fast turn around. The power to push and pull versions will be yours.

**Comment by DamonGant** [comment]
Before you go with snappy, you should be investing in plain old Docker. Wider adaptation, same features.

I think the point is really to use what @JoshData is familiar with.

**Comment by johnjamesjacoby** [comment]
> if we ever make it to 16.04

:(

**Comment by qinwf** [comment]
I use Docker on Ubuntu 16.04 with this, and it works fine.

**Comment by cromulus** [comment]
@qinwf Have you gotten MIAB working in docker? That's been an ongoing struggle, so any insight you might have would be really valuable.

**Comment by qinwf** [comment]
Yes, I am using MIAB in a host with Ubuntu 16.04 with Docker image ubuntu-upstart:14.04.

The mail works fine. 

One thing needs to take care of is that fail2ban does not work inside ubuntu-upstart:14.04. But it works in the host Ubuntu 16.04. I think this issue can be solved, but I did not try to solve it yet.

**Comment by cromulus** [comment]
A bunch of work was done to try to get MIAB running on docker. It was abandoned because it got too complex and didn't work quite right. 

If you have MIAB running in docker, that's awesome. Would you please share your changes back with the community?

Thanks!

**Comment by bvdlingen** [comment]
It might be a better approach to develop and distribute mail-in-a-box using docker. 

See another dockerized mail server for example: https://github.com/andryyy/mailcow-dockerized

**Comment by e-fu** [comment]
what needs to happen for the 16.04/17.04 Upgrade:
Is there a todo list?
Is an switch to Debian an option?

I'm asking because i would like to help.

**Comment by JoshData** [comment]
As I said above, the first thing is a test suite. There's some work on that in #1039 that we should get merged.

**Comment by bvdlingen** [comment]
Wouldn’t it be easier to develop, maintain and deploy Mail-in-a-box on a container? Or containers?

This could have several benefits:

1. Mail-in-a-box developers would have full control over all dependencies
2. upgrades will be easier
3.  users would be able to go with the Linux distribution they are the most familiar with

(not my idea but I have been reading about mailcow, they are now using docker containers for their mailserver, see https://github.com/andryyy/mailcow-dockerized)

**Comment by bronson** [comment]
I'm all for staying on 14.04 until 18.04 drops.  There's only one catch that I can think of: https://help.ubuntu.com/community/UpgradeNotes

So, unless something changes, 14.04->18.04 won't be supported.  It needs to pass through each LTS release: 14.04->16.04->18.04.

Would it be best to do both upgrades at the same time?  Or should they be spread around a bit and done, say, a year apart?

**Comment by bronson** [comment]
Also, as much as I would love love LOVE to move to Debian...  Since it basically requires blowing away and re-provisioning the VM, I'm not sure it would ever be worth the effort.

But I do hope somebody could tell me where I'm wrong.  :)

**Comment by JoshData** [comment]
I don't plan to support in-place upgrades. 

On April 11, 2017 11:00:18 PM EDT, Scott Bronson <notifications@github.com> wrote:
>I'm all for staying on 14.04 until 18.04 drops.  There's only one catch
>that I can think of: https://help.ubuntu.com/community/UpgradeNotes
>
>So, unless something changes, 14.04->18.04 won't be supported.  It
>needs to pass through each LTS release: 14.04->16.04->18.04.
>
>Would it be best to do both upgrades at the same time?  Or should they
>be spread around a bit and done, say, a year apart?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/mail-in-a-box/mailinabox/issues/758#issuecomment-293458741

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by bronson** [comment]
Oh?  So even Fedora/Centos is in bounds?  Let the distro bikeshedding begin!  :)

2018 is going to be an interesting year.  Looking forward to helping out once a direction is chosen.

**Comment by damongant** [comment]
@bronson No - the two things it need to be are

* Currently supported with updates, should be for a while
* Familiar enough to @JoshData so he doesn't have to figure out distro-specific stuff

So I guess that leaves us with Ubuntu LTS! With the test suite progressing MiaB should be easier to port to 18.04 too once that hits.

**Comment by bronson** [comment]
Well, Debian fits those reqs too...  But, agreed, it should be whatever makes @JoshData the happiest (or least miserable anyway).

**Comment by anoma** [comment]
Idea for consideration, aim to officially support LTS but only after the first point release (typically about 2 months later). For example this would mean that during the period between 18.04  and 18.04.1 we would be in beta.

I came to this because I was considering why I wanted to +1 the request for Debian and I was wondering why I cared that much. It quickly occurred to me that my Ubuntu headless experiences are tainted by the volume of patches, especially in the early days, when compared to Debian and that I doubt I would be moving production boxes until the point release was out anyway.

**Comment by pavakpaul** [comment]
What's the status now? Is it available for Ubuntu 16.04? Found [this tutorial](http://idroot.net/linux/install-mail-box-ubuntu-16-04-lts/) for installing Mail-in-a-Box for 16.04!

**Comment by JoshData** [comment]
No progress here. Don't use 16.04.

**Comment by zealot128** [comment]
We are running Mailinabox in (small scale) production now without any (apparent) problems (no OwnCloud, no tested DNS, as we have different solutions for that)

We are running with these adjustments: https://github.com/mail-in-a-box/mailinabox/compare/master...pludoni:ubuntu16.04

**Comment by JoshData** [comment]
Thanks for sharing that. That'll be a huge help when we eventually switch.

(At this point though, I think we'll switch at 18.04 LTS. We can start the preparation for that later this year so that in April we're ready.)

**Comment by pavakpaul** [comment]
Guys help me make the decision then. If I need a mail server now, should I just install 14.04? How can I switch to 18.04 when it is released or 14.04 reaches it's end of life?

**Comment by hachre** [comment]
@pavakpaul If you want to use Mail-in-a-Box, you *have* to use 14.04. Anything else and you're on your own.

**Comment by pavakpaul** [comment]
Ok, got it. Can you please tell me what to do when I want to upgrade? Like is it easy to migrate to 18.04?

**Comment by hachre** [comment]
Since Mail-in-a-Box takes care of clearly separating the data from the OS, it is probably going to be as easy as reinstalling 18.04 and running the Mail-in-a-Box setup again with the user data already present. So about as easy as setting up a new box from scratch, with the added benefit of keeping your mail users, aliases and existing mails.

But this is just a guess of mine - I don't think anyone has officially committed to anything yet.

**Comment by damongant** [comment]
General reminder that user questions should be on [Discourse](https://discourse.mailinabox.email/).

**Comment by divansantana** [comment]
Adding a meaningless vote for a switch to devuan latest stable instead of ubuntu LTS. Would be nice. :grin:

**Comment by jirislav** [comment]
Hey guys, just letting you know, that I've successfully migrated to Ubuntu 16.04 & posted my work to my own fork - see https://github.com/jirislav/mailinabox - I've tested that a few times - it works like a charm ;)

**Comment by damongant** [comment]
The lack of lucene is a little concerning for Roundcube users (and I got a few of those), have you explored any options there?

**Comment by JoshData** [comment]
@jirislav Thanks for sharing this!

**Comment by jirislav** [comment]
@damongant well, someone would have to build the same package as for Ubuntu 14 & put it into the ppa

**Comment by jirislav** [comment]
@damongant or there is other possibility - to run Apache Solr (runs with lucene) and then just plug in the dovecot (I already have Apache Solr automated installation in my fork for [nextant](https://github.com/nextcloud/nextant/) to work)

**Comment by bavicj** [comment]
@jirislav I tried your fork on clean 16.04 LTS, but it looks like the installation did not complete. I am not able to access webadmin interface nor roundcube webmail. These are last messages before installation stopped.

`Installing Postfix (SMTP server)...
Installing Dovecot (IMAP server)...
Creating new user database: /home/user-data/mail/users.sqlite
Installing OpenDKIM/OpenDMARC...
Installing SpamAssassin...
Installing Nginx (web server)...
Installing Roundcube (webmail)...
Installing Nextcloud (contacts/calendar)...
Upgrading to Nextcloud version 12.0.2
Installing solr version 6.6.0`

**Comment by jirislav** [comment]
Oh, I've probably incorporated an installation bug since my last stable version. Try this instead:

```
git clone https://github.com/jirislav/mailinabox.git
cd mailinabox
git checkout v0.23a-ubuntu16-rc1
setup/start.sh
```

**Comment by Extarys** [comment]
Aww, I'm using NextClound under Stretch. I'm looking for a mail server but I would like to avoid usign docker, call me old school.
@jirislav is this "stable" enough?

**Comment by jirislav** [comment]
@Extarys as long as I use this "stable" on my production server, I'll keep it up to date with original mail-in-a-box.

I'm not considering moving to another mailserver solution anytime soon, because I'm satisfied with this

**Comment by Extarys** [comment]
That's enough for me ;) Thanks

**Comment by madera17** [comment]
@jirislav I have set up your fork of MIAB and it is working fine in a Linode with 16.04.3 LTS. I had to use the details you posted on 11th Sept as I had the same problem as @bavicj with the most recent release.

Just one question from a newbie. Now it is up and running what is the update procedure to keep it up to date? Thanks.

**Comment by jirislav** [comment]
@madera17 Update procedure is following:
- Check out [new releases](https://github.com/jirislav/mailinabox/tags)
- if there is tag with newer version than you have (it has to contain `ubuntu16` in it's name), then execute these commands to upgrade:
```bash
cd mailinabox
git checkout NAME_OF_THE_TAG_HERE
setup/start.sh
```

**Comment by madera17** [comment]
@jirislav Thank you for the info.
Unfortunately when attempting to upgrade 'git checkout v0.24-ubuntu16' after running 'setup/start.sh' it ends as described by @bavicj with the last line -

Installing solr version 6.6.0`

Setup appears not to complete.

**Comment by jirislav** [comment]
@madera17 Oh, I'm sorry for that inconvenience with the Solr :( 

I plan to move the NextAnt + Solr upgrades to a separate module, which you will have the possibility to choose whether you want to install it or not, so that the main installation routine `setup/start.sh` stays focused on preparing clean mailserver solution.

**Comment by Memphizzz** [comment]
@jirislav would you be able to replicate these changes on a 18.04 installation a few months from now? If so, wouldn't this be a good point to have MIAB switch to 18.04 as @JoshData suggested approx. 2 years ago?

**Comment by jirislav** [comment]
@Memphizzz I actually think it is already compatible with 18.04 :) .. I just have to try it out & fix problems that appear

**Comment by Memphizzz** [comment]
@jirislav Sounds good but keep in mind, 18.04 ain't released yet. AFAIK the feature freeze didn't even happen yet, so stuff might still change until the end of april.

**Comment by ppKrauss** [comment]
About install with *Ubuntu 16 LTS 64-bit*, some news?  Can I use in it? ... Another option is to use a minimal hardware (ex. [DigitalOcean droplet](https://www.digitalocean.com/pricing/#droplet) of 512Mb RAM 20 Gb HD and 1vCPU) with Ubuntu 14... There are some  "minimal configuration" tips here? (no at README or site).

**Comment by ilyaevseev** [comment]
Really???

    Mail-in-a-Box only supports being installed on Ubuntu 14.04, sorry. You are running:

    Ubuntu 16.04 LTS

    We can't write scripts that run on every possible setup, sorry.

**Comment by jirislav** [comment]
@ilyaevseev really what?

**Comment by cybercode** [comment]
According to the ubuntu CVE on the meltdown exploit, 14.x won't be updated because it is EOL. It is of course possible there will be a backport, but... Any plans to address this issue?

**Comment by yodax** [comment]
Do you have a link to that statement? 

[This](http://people.canonical.com/~ubuntu-security/cve/main.html) shows it at pending. [This](https://wiki.ubuntu.com/SecurityTeam/KnowledgeBase/SpectreAndMeltdown) doesn't mention 14.04 as not supported. 14.04 isn't EOL as far as I know.

**Comment by cybercode** [comment]
Sorry, think i mis-read this page: https://people.canonical.com/%7Eubuntu-security/cve/2017/CVE-2017-5754.html,
see the second entry below...

Package
Source: linux-lts-trusty (LP Ubuntu Debian)
Upstream:
needed
Ubuntu 12.04 ESM (Precise Pangolin):
pending
Ubuntu 14.04 LTS (Trusty Tahr):
DNE
Ubuntu 16.04 LTS (Xenial Xerus):
DNE
Ubuntu 17.04 (Zesty Zapus):
DNE
Ubuntu 17.10 (Artful Aardvark):
DNE
Ubuntu 18.04 LTS (Bionic Beaver):
DNE
Package
Source: linux-lts-wily (LP Ubuntu Debian)
Upstream:
needed
Ubuntu 12.04 ESM (Precise Pangolin):
DNE
Ubuntu 14.04 LTS (Trusty Tahr):
ignored (end-of-life)
Ubuntu 16.04 LTS (Xenial Xerus):
DNE
Ubuntu 17.04 (Zesty Zapus):
DNE
Ubuntu 17.10 (Artful Aardvark):
DNE
Ubuntu 18.04 LTS (Bionic Beaver):
DNE

**Comment by JoshData** [comment]
Folks, please open a new issue to talk about a new security issue.

**Comment by patricmutwiri** [comment]
Do we have mail-in-a-box for ubuntu 16.04?

**Comment by Memphizzz** [comment]
@patricmutwiri No, but as @jirislav pointed out, we might get a version for 18.04 once its released.

**Comment by patricmutwiri** [comment]
Let me just head to iRedmail

**Comment by nikolay** [comment]
There are a lot of 404s via `apt-get update && apt-get upgrade`. Any suggestion on the `sources.list` and `sources.list.d` contents?

**Comment by anoma** [comment]
No 404s here. Suggest you create a dedicated ticket unrelated to this one.

**Comment by JoshData** [comment]
Closing in favor of #1358.

**Comment by CHazz** [comment]
> Ugh has it been two years already?
> 
> I'm tempted to wait two more years and switch at what would presumably be 18.04 LTS....

any news ? 14.04 is ,,, little retro :) maybe 19.04 ?

**Comment by JoshData** [comment]
See #1358.
